rflx/specification/integration.py

```diff
@@ -0,0 +1,137 @@
+            return
+        for var_name in integration.buffer_size.global_.keys():
+            if var_name not in [x.parts[-1] for x in session.declarations.keys()]:
+                fail(
+                    (
+                        f'{filename}: unknown variable "{var_name}" in global state '
+                        f'of session "{session.identifier.parts[-1]}"'
+                    ),
+                    Subsystem.PARSER,
+                    Severity.ERROR,
+                    None,
+                )
+
+    @staticmethod
+    def __validate_states(
+        filename: Path, integration: SessionIntegration, session: Session
+    ) -> None:
+        if integration.buffer_size.local_ is None:
+            return
+        for state_name, state_entry in integration.buffer_size.local_.items():
+            state = None
+            for s in session.states:
+                if s.identifier.parts[-1] == state_name:
+                    state = s
+            if state is None:
+                fail(
+                    (
+                        f'{filename}: unknown state "{state_name}" in session '
+                        f'"{session.identifier.parts[-1]}"'
+                    ),
+                    Subsystem.PARSER,
+                    Severity.ERROR,
+                    None,
+                )
+            for var_name in state_entry:
+                if var_name not in [x.parts[-1] for x in state.declarations.keys()]:
+                    fail(
+                        (
+                            f'{filename}: unknown variable "{var_name}" in state '
+                            f'"{state_name}" of session "{session.identifier.parts[-1]}"'
+                        ),
+                        Subsystem.PARSER,
+                        Severity.ERROR,
+                        None,
+                    )
+
+    def validate(self, model: Model) -> None:
+        for filename, file in self._files.items():
+            for session_name, integration in file.session.items():
+                session = None
+                for s in model.sessions:
+                    if s.identifier.parts[-1] == session_name:
+                        session = s
+                if (
+                    session is None
+                    or session.location is None
+                    or session.location.source is None
+                    or session.location.source.with_suffix(".rfi") != filename
+                ):
+                    fail(
+                        f'{filename}: unknown session "{session_name}"',
+                        Subsystem.PARSER,
+                        Severity.ERROR,
+                        None,
+                    )
+                self.__validate_globals(filename, integration, session)
+                self.__validate_states(filename, integration, session)
+
+    def get_size(self, session: Session, state: Optional[State], var: ID) -> int:
+        assert session.location is not None
```

Comment:
Only the state identifier should be used here. In general, having the possibility to pass a `state` which is potentially not part of `session` could lead to strange behavior. In the particular case, only the identifier of the state is used, so there is not even the need for the whole state object.

Reply:
fixed in 3c684cf25a88b85b6cf8664ddc5096cd6e4bc2a0